### PR TITLE
[ignition-plugin1] Add new port 🤖

### PR DESCRIPTION
**Pull request by traversaro**
- What does your PR fix? 
  - This PR adds a new port for the new major  version of the [ignition-plugin](https://ignitionrobotics.org/libs/plugin) library. 

- Which triplets are supported/not supported?
  - All tested triplet should be supported

- Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?
  - I hope.

As discussed in https://github.com/microsoft/vcpkg/pull/7781, different major version of ignition robotics libraries (https://ignitionrobotics.org/) can be installed side by side, so each new major version is added as a new port. 

In particular, ignition-plugin1 is a dependency of Ignition Robotics Citadel (that contains Ignition Gazebo 3, see https://www.openrobotics.org/blog/2019/12/11/ignition-citadel-released).

### CODE REVIEW

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
@strega-nil
Could you please help handle the failure on osx pipeline?
`##[error]Git checkout failed with exit code: 128`

**Comment by strega-nil** [comment]
These seem to be real errors, but I don't understand them.

**Comment by NancyLi1013** [comment]
@strega-nil 
Thanks for your help.
The regressions on x64-osx seem like this:

```
-- Checking file: /Volumes/data/work/1/s/packages/ignition-plugin1_x64-osx/lib/pkgconfig/ignition-plugin1-loader.pc
CMake Error at scripts/cmake/vcpkg_fixup_pkgconfig.cmake:130 (message):
  Library dl was not found! If it is a system library use the
  SYSTEM_LIBRARIES parameter for the vcpkg_fixup_pkgconfig call! Otherwise,
  corret the *.pc file
Call Stack (most recent call first):
  scripts/cmake/vcpkg_fixup_pkgconfig.cmake:232 (vcpkg_fixup_pkgconfig_check_libraries)
  installed/x64-osx/share/ignitionmodularscripts/ignition_modular_library.cmake:29 (vcpkg_fixup_pkgconfig)
  installed/x64-osx/share/ignitionmodularscripts/ignition_modular_library.cmake:131 (ignition_modular_build_library)
  ports/ignition-plugin1/portfile.cmake:4 (ignition_modular_library)
  scripts/ports.cmake:90 (include)
```
@traversaro 
Could you please take a look about this?

**Comment by traversaro** [comment]
Probably it is necessary to pass `dl` as `SYSTEM_LIBRARIES` of `vcpkg_fixup_pkgconfig`, but before we need to expose it in `ignition_modular_library`, I will  prepare a  commit to address that.

**Comment by traversaro** [comment]
> 
> 
> Probably it is necessary to pass `dl` as `SYSTEM_LIBRARIES` of `vcpkg_fixup_pkgconfig`, but before we need to expose it in `ignition_modular_library`, I will prepare a commit to address that.

Actually `dl` is already included in the system libraries in Linux (see https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_fixup_pkgconfig.cmake#L195),  probably it make sense to do so  also on the other platforms that support  the dl library.

**Comment by NancyLi1013** [comment]
It seems that `dl ` is also required to the system libraries on osx?

@Neumann-A 
Could you please help comment this?

**Comment by traversaro** [comment]
> 
> 
> It seems that `dl ` is also required to the system libraries on osx?
> 
> @Neumann-A
> Could you please help comment this?

I investigate a bit and indeed it make sense to have a `libdl` as system library on macOS, but on the other hand on macOS you don't have `libm`. 
I will update this in the weekend, but basically I was thinking of  changing the lines (from https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_fixup_pkgconfig.cmake#L195):
~~~
    if(VCPKG_TARGET_IS_LINUX)
        list(APPEND _vfpkg_SYSTEM_LIBRARIES -ldl -lm)
    endif()
~~~
to 
~~~
    # Platforms with libdl
    if(VCPKG_TARGET_IS_LINUX OR VCPKG_TARGET_IS_OSX OR VCPKG_TARGET_IS_ANDROID)
        list(APPEND _vfpkg_SYSTEM_LIBRARIES -ldl)
    endif()
    # Platforms with libm
    if(VCPKG_TARGET_IS_LINUX OR VCPKG_TARGET_IS_FREEBSD OR VCPKG_TARGET_IS_ANDROID)
        list(APPEND _vfpkg_SYSTEM_LIBRARIES -lm)
    endif()
~~~
Other  platforms used in community triplets probably will also need to be added to this if, but for now I only added the system considered in https://github.com/microsoft/vcpkg/blob/master/scripts/cmake/vcpkg_common_definitions.cmake .

**Comment by Neumann-A** [comment]
@traversaro:

You can either do that or call `vcpkg_fixup_pkgconfig` with `SYSTEM_LIBRARIES`. E.g. `vcpkg_fixup_pkgconfig(SYSTEM_LIBRARIES <cmake list of libraries>)`

Probably also needs a case for `VCPKG_TARGET_IS_MINGW`. Maybe it makes more sense to define a global variable called `VCPKG_SYSTEM_LIBRARIES` in `vcpkg_common_definitions`? This would allow them to be overwritten globally from the triplet.

**Comment by traversaro** [comment]
> Maybe it makes more sense to define a global variable called `VCPKG_SYSTEM_LIBRARIES` in `vcpkg_common_definitions`? This would allow them to be overwritten globally from the triplet.

That make totally sense.

**Comment by traversaro** [comment]
> Probably also needs a case for `VCPKG_TARGET_IS_MINGW`.

For  which library? As far as I know on MinGW dl is provided by an external library (dlfcn-win32), but indeed I do not know if linking with  `libm` is necessary or not on MinGW.

**Comment by Neumann-A** [comment]
>For which library? 

don't know mingw or msys that well. Just thought that there might be a need for an additional case for it as well ;)

**Comment by traversaro** [comment]
@NancyLi1013 @Neumann-A 
The PR has been updated to handle system libraries in a more general way, see https://github.com/microsoft/vcpkg/pull/11275/commits/f38a8cc6546da728153642c750743977ea167bf1 . This should  fix the CI failure on macOS . 

@NancyLi1013 Thanks to the  comment of @Electric-Turtle in https://github.com/microsoft/vcpkg/pull/11273#issuecomment-628832899 I noticed a critical flaw in this  PR. Namely, `ignition-plugin1` install multiples CMake  config files,  and it was necessary to modify vcpkg_fixup_cmake_targets  (https://github.com/microsoft/vcpkg/pull/11275/commits/4988fe6f228b3659cc25907184b542fd28b51d71) and  ignition_modular_library (https://github.com/microsoft/vcpkg/pull/11275/commits/b587d0fb41a07dce83ecd6c52332dc4f539fb2b9) to handle this case.

**Comment by NancyLi1013** [comment]
The recent changes seems cause many regressions. 
@traversaro 
Could you please look into them?

**Comment by traversaro** [comment]
> 
> 
> The recent changes caused some new regressions.
> @traversaro
> Could you please look into them?

I will check them, but those seems unrelated failures that  appear here because `vcpkg_fixup_cmake_targets` has been modified, see https://github.com/microsoft/vcpkg/pull/11502 where a small space modification in `vcpkg_fixup_cmake_targets` caused exactly the same regressions.

**Comment by traversaro** [comment]
> 
> 
> > The recent changes caused some new regressions.
> > @traversaro
> > Could you please look into them?
> 
> I will check them, but those seems unrelated failures that appear here because `vcpkg_fixup_cmake_targets` has been modified, see #11502 where a small space modification in `vcpkg_fixup_cmake_targets` (that does not change anything in practice) caused exactly the same regressions.

I checked the error messages, and it seems that the port are failing due to the MSYS2 regressions that is tracked and discussed in several issues and PR, in particular https://github.com/microsoft/vcpkg/pull/11471 .

**Comment by traversaro** [comment]
I may be wrong, but the CI failures seems not to be related with this PR.

**Comment by strega-nil** [comment]
@traversaro I'll look into this tomorrow

**Comment by traversaro** [comment]
PR rebased over latest master.

**Comment by NancyLi1013** [comment]
All checks seem to be successful.

File: scripts/cmake/vcpkg_fixup_pkgconfig.cmake [hunk 1]
```diff
@@ -194,8 +194,8 @@ endfunction()
 function(vcpkg_fixup_pkgconfig)
     cmake_parse_arguments(_vfpkg "" "" "RELEASE_FILES;DEBUG_FILES;SYSTEM_LIBRARIES;SYSTEM_PACKAGES;IGNORE_FLAGS" ${ARGN})
     
-    if(VCPKG_TARGET_IS_LINUX)
-        list(APPEND _vfpkg_SYSTEM_LIBRARIES -ldl -lm)
+    if(VCPKG_SYSTEM_LIBRARIES)
```
Comment (dan-shaw) [inline scripts/cmake/vcpkg_fixup_pkgconfig.cmake:197]: Does `pkgconfig` need the system libraries on Windows?

**Review by dan-shaw — APPROVED** [review]

File: scripts/cmake/vcpkg_fixup_pkgconfig.cmake (hunk 1 above)
Comment (traversaro) [reply scripts/cmake/vcpkg_fixup_pkgconfig.cmake:197]: Yes, the system libraries that I have encounter in pkg-config files on Windows are defined  in https://github.com/microsoft/vcpkg/pull/11275/files#diff-c1846c91b47ff94ccee5e8a6a1658e62R130 . In  particular, I found them in the pcap port (see https://github.com/microsoft/vcpkg/blob/2020.06/ports/libpcap/portfile.cmake#L89), and adding them in a centralized place seems to me consistent to what we are doing in Linux/macOS .

**Comment by dan-shaw** [comment]
Thanks for the PR!

**Pull request merged**